xla/hlo/transforms/collectives/collective_quantizer.cc

```diff
@@ -97,22 +97,26 @@ HloInstruction* ApplyUnaries(HloInstruction* instr,
   return instr;
 }
 
-// Returns whether instr is replicated across all partitions associated with
-// module. Returns false when there are multiple replicas.
-absl::StatusOr<bool> InstrIsReplicated(HloModule* module,
-                                       HloInstruction* instr) {
+// Returns whether instr is replicated across replica_groups. When
+// replica_groups is empty, returns whether instr is replicated across all
+// partitions associated with module. Returns false when there are multiple
+// replicas.
+absl::StatusOr<bool> InstrIsReplicated(
+    HloModule* module, HloInstruction* instr,
+    absl::Span<const ReplicaGroup> replica_groups) {
   // The replication analysis only verifies the replication of instr across
   // partitions, not replicas. The replica count must be one to ensure instr is
   // replicated across all devices.
   if (module->config().replica_count() > 1) {
     return false;
   }
 
-  TF_ASSIGN_OR_RETURN(
-      auto replication_analysis,
-      HloReplicationAnalysis::Run(module,
-                                  /*cross_partition_spmd=*/true));
-  return replication_analysis->HloInstructionIsReplicatedAt(instr, {});
+  TF_ASSIGN_OR_RETURN(auto replication_analysis,
+                      HloReplicationAnalysis::RunWithPartialReplication(
+                          module,
+                          /*cross_partition_spmd=*/true));
+  return replication_analysis->HloInstructionIsReplicatedAt(instr, {},
```

Comment:
This is not very efficient.
You can run the replication analysis once as you don't remove instructions during rewrite, you can query the same analysis result.